core/src/main/java/com/datastax/oss/driver/api/core/uuid/Uuids.java

```diff
@@ -19,6 +19,7 @@
 import com.datastax.oss.driver.internal.core.util.Loggers;
```

Comment:
Changes in his file are an attempt to handle the UnsatisfiedLinkErrors reported elsewhere on this PR.  By deferring initialization of state necessary to support time-based UUIDs until run-time we prevent the build-time analysis from detecting any attempt to use the linked LLVM native code.  The drawback here is that we've replaced a simple field access with a method call to a Supplier plus some (at the moment manual) unboxing into a primitive, so there is some modest performance degradation here.

@olim7t are you okay with this?

Reply:
Additional note: I originally tried to implement this via some Graal substitutions for the underlying methods in Uuids, but the complexity of that approach escalated quickly.  You'd want to delete the field performing static initialization and inject a new field which uses the supplier-based approach here, but you can't inject static fields.  You also want to preserve access to the underlying functionality (to use in your supplier) without copying code... which proves to be difficult/impossible to do.

Reply:
I think it is possible to substitute static fields, see `@Alias` annotation and it's usages in the `quarkus` repo

Reply:
I'm okay on the principle, but I agree that introducing a supplier and unboxing is unfortunate. If all we need is execution on first actual use, I think the lazy initialization holder idiom will work as well:
```java
public final class Uuids {
  private static class ClockSeqAndNodeHolder {
    private static final long CLOCK_SEQ_AND_NODE = makeClockSeqAndNode();

    // not strictly necessary, but it feels cleaner to move this method into the holder as well:
    private static long makeClockSeqAndNode() { ... }
  }
  ...
}
```

Reply:
Weirdly that actually _doesn't_ work in this case @olim7t.  That pattern relies on the lazy initialization of classes (i.e. not initialized until loaded), but the problem here is that the class in question (api.core.uuid.Uuids) is being initialized at build time.  As a result the loader code is exec'd immediately, voiding any benefit.

To validate, I implemented the lazy init holder pattern via the following diff: 

```diff
diff --git a/core/src/main/java/com/datastax/oss/driver/api/core/uuid/Uuids.java b/core/src/main/java/com/datastax/oss/driver/api/core/uuid/Uuids.java
index 0d633df15..d95146b23 100644
--- a/core/src/main/java/com/datastax/oss/driver/api/core/uuid/Uuids.java
+++ b/core/src/main/java/com/datastax/oss/driver/api/core/uuid/Uuids.java
@@ -19,7 +19,6 @@ import com.datastax.oss.driver.internal.core.os.Native;
 import com.datastax.oss.driver.internal.core.util.Loggers;
 import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
 import com.datastax.oss.driver.shaded.guava.common.base.Charsets;
-import com.datastax.oss.driver.shaded.guava.common.base.Suppliers;
 import edu.umd.cs.findbugs.annotations.NonNull;
 import java.lang.management.ManagementFactory;
 import java.net.InetAddress;
@@ -40,7 +39,6 @@ import java.util.Set;
 import java.util.TimeZone;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicLong;
-import java.util.function.Supplier;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -115,8 +113,17 @@ public final class Uuids {
   private Uuids() {}
 
   private static final long START_EPOCH = makeEpoch();
-  private static final Supplier<Long> CLOCK_SEQ_AND_NODE =
-      Suppliers.memoize(Uuids::makeClockSeqAndNode);
+
+  private static class ClockSeqAndNodeContainer {
+
+    private static final long VALUE = makeClockSeqAndNode();
+
+    private synchronized long get() {
+      return VALUE;
+    }
+  }
+
+  private static final ClockSeqAndNodeContainer CLOCK_SEQ_AND_NODE = new ClockSeqAndNodeContainer();
 
   // The min and max possible lsb for a UUID.
   //
@@ -440,7 +447,7 @@ public final class Uuids {
    */
   @NonNull
   public static UUID timeBased() {
-    return new UUID(makeMsb(getCurrentTimestamp()), CLOCK_SEQ_AND_NODE.get().longValue());
+    return new UUID(makeMsb(getCurrentTimestamp()), CLOCK_SEQ_AND_NODE.get());
   }
 
   /**
```

This worked exactly as expected when running the app directly via a JVM (no Graal):

```
817 [s0-admin-0] DEBUG com.datastax.oss.driver.internal.core.util.Reflection  - Trying with default package com.datastax.oss.driver.internal.core.time.AtomicTimestampGenerator
1337 [s0-admin-0] INFO com.datastax.oss.driver.internal.core.time.Clock  - Using native clock for microsecond precision
...
1631 [s0-io-2] DEBUG com.datastax.oss.driver.internal.core.channel.ProtocolInitHandler  - [s0|control|id: 0x5da58562, L:/127.0.0.1:44588 - R:localhost/127.0.0.1:9042] step OPTION
S received response opcode=SUPPORTED                                                                                                                                              
1657 [s0-io-2] INFO com.datastax.oss.driver.api.core.uuid.Uuids  - PID obtained through native call to getpid(): 23083                                                            
1662 [s0-io-2] DEBUG com.datastax.oss.driver.internal.core.channel.InFlightHandler  - [s0|control|id: 0x5da58562, L:/127.0.0.1:44588 - R:localhost/127.0.0.1:9042] Got last respon
se on in-flight stream id 0, completing and releasing  
```

and when building and running my local keyspace counting app via Graal (no Quarkus):

```
11 [s0-admin-0] DEBUG com.datastax.oss.driver.internal.core.util.Reflection  - Trying with default package com.datastax.oss.driver.internal.core.time.AtomicTimestampGenerator    
11 [s0-admin-0] INFO com.datastax.oss.driver.internal.core.os.Native  - Using Graal-specific native functions                                                                     
11 [s0-admin-0] INFO com.datastax.oss.driver.internal.core.time.Clock  - Using native clock for microsecond precision   
...
46 [s0-io-2] DEBUG com.datastax.oss.driver.internal.core.channel.ProtocolInitHandler  - [s0|control|id: 0xcd094fd9, L:/127.0.0.1:44638 - R:localhost/127.0.0.1:9042] step OPTIONS
received response opcode=SUPPORTED
47 [s0-io-2] INFO com.datastax.oss.driver.api.core.uuid.Uuids  - PID obtained through native call to getpid(): 23957
48 [s0-io-2] DEBUG com.datastax.oss.driver.internal.core.channel.InFlightHandler  - [s0|control|id: 0xcd094fd9, L:/127.0.0.1:44638 - R:localhost/127.0.0.1:9042] Got last response
 on in-flight stream id 0, completing and releasing
```

Note that in each of these cases UUID initialization is clearly deferred as expected.  Things fall apart when I try to build a Quarkus app, though.  We already know that Quarkus is somehow triggering build-time init of the Uuids class, and that eventually leads me here:

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /work/local/graal-19.3.1/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.
ch.maxUpdateArraySize=100 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDe
tection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.g
enscavenge.CollectionPolicy$BySpaceAndTime -H:+JNI -jar quickstart-1.0.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:-IncludeA
llTimeZones -H:EnableURLProtocols=http -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quickstart-1.0.0-SNAPSHOT-runner                     
[quickstart-1.0.0-SNAPSHOT-runner:24275]    classlist:  24,108.06 ms                                                                                                             
[quickstart-1.0.0-SNAPSHOT-runner:24275]        (cap):   1,713.83 ms                                                                                                             
[quickstart-1.0.0-SNAPSHOT-runner:24275]        setup:   5,122.67 ms
10:19:30,365 INFO  [com.dat.oss.dri.int.cor.DefaultMavenCoordinates] DataStax Java driver for Apache Cassandra(R) (com.datastax.oss:java-driver-core) version 4.7.0-SNAPSHOT
10:19:33,420 INFO  [org.jbo.threads] JBoss Threads version 3.0.1.Final              
10:19:58,203 INFO  [com.dat.oss.dri.int.cor.os.Native] Using Graal-specific native functions                                                                                     
10:19:58,207 INFO  [com.dat.oss.dri.api.cor.uui.Uuids] PID obtained through JMX: 24275 
[quickstart-1.0.0-SNAPSHOT-runner:26961]   (typeflow):  60,717.82 ms
```

Note the initialization of Uuids at build-time here.  This disappears entirely when I rollback the changes in the diff and go back to the Supplier-based impl currently in the PR:

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /work/local/graal-19.3.1/bin/native-image -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jb
oss.logmanager.LogManager -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -H:+JNI -jar quickstart-1.0.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:-IncludeAllTimeZones -H:EnableURLProtocols=http -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quickstart-1.0.0-SNAPSHOT-runner
[quickstart-1.0.0-SNAPSHOT-runner:26961]    classlist:  20,907.58 ms
[quickstart-1.0.0-SNAPSHOT-runner:26961]        (cap):   1,187.25 ms
[quickstart-1.0.0-SNAPSHOT-runner:26961]        setup:   3,987.47 ms
10:42:59,570 INFO  [com.dat.oss.dri.int.cor.DefaultMavenCoordinates] DataStax Java driver for Apache Cassandra(R) (com.datastax.oss:java-driver-core) version 4.7.0-SNAPSHOT
10:43:01,875 INFO  [org.jbo.threads] JBoss Threads version 3.0.1.Final
10:43:32,368 INFO  [com.dat.oss.dri.int.cor.os.Native] Using Graal-specific native functions
[quickstart-1.0.0-SNAPSHOT-runner:26961]   (typeflow):  60,717.82 ms
```

Note that lack of Uuids initialization here

It _is_ true that the Quarkus build case does succeed now rather than throwing the UnsatisfiedLinkError as it did before.  That's because the recent changes to this PR cause GraalLibc.available() to return true iff it's able to execute a getpid() and gettimeofday() native call without errors or exceptions.  Those calls fail in the Quarkus build case (due to the UnsatisfiedLinkError discussed above) so Native is "unavailable" in this context.  The initialization logic of Uuids knows to try JMX next in that case and that succeeds.  So there's a semi-plausible argument that we just live with this constraint and move on.

Note, however, that there's a built-in constraint here; we'd now be saying that _no_ logic which might get executed build-time could use os.Native without some fallback option (like JMX for getting the PID in this case).  That may be okay but it is worth highlighting.

Reply:
Follow-up: I'm pretty sure I can adapt the memoized Supplier code to avoid the Supplier dependency and (un)boxing issues... working on that now.